buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/BuildForkPointDistribution.java

```diff
@@ -0,0 +1,121 @@
+import java.nio.file.Files;
+import java.util.stream.Stream;
+
+@CacheableTask
+public class BuildForkPointDistribution extends DefaultTask {
+    private static final Logger LOGGER = Logging.getLogger(BuildForkPointDistribution.class);
+
+    private String forkPointCommitId;
+
+    private String baselineVersion = "defaults";
+
+    @Input
+    @Optional
+    public String getForkPointCommitId() {
+        return forkPointCommitId;
+    }
+
+    public void setForkPointCommitId(String forkPointCommitId) {
+        this.forkPointCommitId = forkPointCommitId;
+    }
+
+    @OutputDirectory
+    public File getForkPointDistributionDir() {
+        return new File(getProject().getRootProject().getBuildDir(), "distributions/gradle-forkpoint");
+    }
+
+    @Internal
+    public String getBaselineVersion() {
+        return baselineVersion;
+    }
+
+    @TaskAction
+    void buildDistribution() {
+        LOGGER.quiet("Building fork point distribution for: " + forkPointCommitId);
+        try {
+            prepareGradleRepository();
+            tryBuildDistribution(forkPointCommitId);
+            setBaselineVersion();
+            LOGGER.quiet("Building commit " + forkPointCommitId + " succeeded, now the baseline is " + baselineVersion);
+        } catch (Exception e) {
+            LOGGER.quiet("Building commit " + forkPointCommitId + " failed, fallback to default baseline.", e);
+        }
+    }
+
+    private void setBaselineVersion() {
+        baselineVersion = Stream.of(getForkPointDistributionDir().listFiles())
+            .filter(file -> file.getName().endsWith("-bin.zip"))
+            // gradle-5.0-commit-2149a1d-bin.zip -> 5.0-commit-2149a1d
+            .map(file -> file.getName().substring("gradle-".length(), file.getName().length() - "-bin.zip".length())).findFirst()
+            .get();
+    }
+
+    private void prepareGradleRepository() throws IOException {
+        File cloneDir = getGradleCloneTmpDir();
+        if (new File(cloneDir, ".git").isDirectory()) {
+            exec(cloneDir, "git", "reset", "--hard");
+            exec(cloneDir, "git", "fetch");
+        } else {
+            FileUtils.forceMkdir(cloneDir);
+            exec(cloneDir.getParentFile(), "git", "clone", getProject().getRootDir().getAbsolutePath(), getGradleCloneTmpDir().getAbsolutePath(), "--no-checkout");
+        }
+    }
+
+    private void exec(File workingDir, Object... command) {
+        getProject().exec(spec -> {
+            spec.commandLine(command);
+            spec.workingDir(workingDir);
+        });
+    }
+
+    private File getGradleCloneTmpDir() {
+        return new File(getProject().getRootProject().getBuildDir(), "tmp/gradle-find-forkpoint");
+    }
+
+
+    private void tryBuildDistribution(String commit) throws Exception {
+        exec(getGradleCloneTmpDir(), "git", "checkout", commit, "--force");
+        exec(getGradleCloneTmpDir(), "./gradlew", ":distributions:binZip", ":toolingApi:toolingApiShadedJar");
```

Comment:
We are running `clean` on CI since we are not sure enough what the effect of leftovers is. Especially, when we store something in the cache. I think we should do that here, too.

Reply:
Done.